beaker/precompile.py

```diff
@@ -50,26 +55,26 @@ class PrecompileTemplateValue:
     pc: int = 0
 
 
-class Precompile:
+class Precompile(ABC):
     """
-    Precompile allows a contract to signal that some LogicSignature should be
-    fully compiled prior to trying to construct the approval and clear programs.
+    Precompile allows a smart contract to signal that some child contract should be
+    fully compiled prior to constructing its own program. This is the parent class of
+    AppPrecompile and LSigPrecompile and should not be instantiated directly.
     """
 
-    def __init__(self, teal_src: str | None):
-
-        if teal_src is None:
-            raise TealInputError("teal_src cannot be None")
-
-        self.program = teal_src
+    @abstractmethod
+    def __init__(self):
+        super().__init__()
 
-        self.binary: Optional[bytes] = None
-        self.program_hash: Optional[str] = None
-        self.map: Optional[SourceMap] = None
+        self.smart_contract: "Application | LogicSignature" = None
```

Comment:
I don't love the new term here `smart_contract` but I don't have a better suggestion initially

Reply:
Maybe we flip this? That is, make PrecompileApp/PrecompileLSig hold Precompiles (2 for apps, 1 for lsig) then i think the API can be standard and no need for abstract stuff. 

I think we'd just need to tweak the init of the Application class to look for either the PrecpompileApp or PrecompileLSig here: https://github.com/algorand-devrel/beaker/blob/0667d68c6c7cc8886d091f25afaec9d0fd034475/beaker/application.py#L126

Reply:
Did the flip, let me know if it looks better to you.